### PR TITLE
feat: add support for testing S3 API locally 

### PR DESCRIPTION
**Pull request by akshbhu**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,10 +6,15 @@ coverage
...
 .vscode
 packages/graphql*/lib
+packages/amplify-storage-simulator/lib
+packages/amplify-storage-simulator/node_modules
 packages/amplify-graphql*/lib
 packages/amplify-e2e-tests/lib
 packages/*/node_modules
 packages/*/package-lock.json
+packages/amplify-util-mock/lib
+packages/amplify-util-mock/node_modules
```
Comment (yuth) [inline .gitignore]: This is not needed as we have a general rule on line#16

File: packages/amplify-frontend-android/lib/frontend-config-creator.js [hunk 2]
```diff
@@ -43,6 +43,9 @@ function getAWSConfigObject(amplifyResources) {
     },
   };
 
+  if (amplifyResources.testMode) {
+    configOutput.DangerouslyConnectToHTTPEndpointForTesting = true;
```
Comment (yuth) [inline packages/amplify-frontend-android/lib/frontend-config-creator.js]: Is this needed? We decided not to have a global flag

File: packages/amplify-storage-simulator/package.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  "description": "An S3 simulator to test S3 APIs",
+  "main": "lib/index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "tsc",
+    "watch": "tsc -w",
+    "start": "node lib/test/test.js"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: We should have the same license as other packages. `Apache-2.0` and author should be `Amazon Web Services`

File: packages/amplify-storage-simulator/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  "dependencies": {
+    "@hapi/topo": "^3.1.2",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: Do you need this?

File: packages/amplify-storage-simulator/package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    "@types/body-parser": "^1.17.0",
+    "@types/cors": "^2.8.5",
+    "@types/glob": "^7.1.1",
+    "@types/serve-static": "^1.13.2",
+    "body-parser": "^1.19.0",
+    "chokidar": "^3.0.2",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: Do you need this?

File: packages/amplify-storage-simulator/package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    "event-to-promise": "^0.8.0",
+    "express": "^4.17.1",
+    "glob": "^7.1.4",
+    "ip": "^1.1.5",
+    "lodash": "^4.17.14",
+    "md5-file": "^4.0.0",
+    "object-to-xml": "^2.0.0",
+    "save": "^2.4.0",
+    "serve-static": "^1.14.1",
+    "tsc-watch": "^2.2.1",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: and this?

File: packages/amplify-storage-simulator/package.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    "typescript": "^3.5.3",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: these can go to `devDependencies`

File: packages/amplify-storage-simulator/package.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    "typescript-logging": "^0.6.3",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: these can go to devDependencies

File: packages/amplify-storage-simulator/package.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "tsc",
+    "watch": "tsc -w",
+    "start": "node lib/test/test.js"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "@hapi/topo": "^3.1.2",
+    "@types/body-parser": "^1.17.0",
```
Comment (yuth) [inline packages/amplify-storage-simulator/package.json]: All the `@types/*` packages should be devDependencies

File: packages/amplify-storage-simulator/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import {StorageSimulatorServer} from './server';
+
+export interface StorageSimulatorDataSourceBaseConfig {
+  name: string;
+  type: string;
+}
+
+export type StorageSimulatorServerConfig = {
+  port: number;
+  wsPort: number;
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/index.ts]: Why do you need `wsPort`

File: packages/amplify-storage-simulator/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import {StorageSimulatorServer} from './server';
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/index.ts]: Why can't this logic go to `StorageSimulatorServer `

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 12]
```diff
@@ -0,0 +1,327 @@
+import * as express from 'express';
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: Could you format your files using prettier or something similar

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+
+  constructor(
+    private config: StorageSimulatorServerConfig,
+  ) {
+    this.localDirectoryPath = config.localDirS3;
+    //console.log("path file", this.localDirectoryPath);
+    this.app = express();
+    this.app.use(express.json());
+    this.app.use(cors(corsOptions));
+    //this.app.set('etag', false);
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: Remove the commented code unless you are documenting something

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+    //this.app.use('/', express.static(STATIC_ROOT))
+    this.app.use(bodyParser.raw({ limit: '100mb', type: '*/*' }));
+    this.app.use(bodyParser.json({ limit: '50mb', type: '*/*' }));
+    this.app.use(bodyParser.urlencoded({ limit: '50mb', extended: false, type: '*/*' }));
+    this.app.use(serveStatic(this.localDirectoryPath), this.handleRequestAll.bind(this));
+
+    this.server = null;
+    this.route = config.route;
+    //console.log("config object = ", config);
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: same here

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+        }
+      }));
+    }
+  }
+
+  private async handleRequestList(request, response) {
+    // fill in  this content
+    // console.log("enter list");
+    let ListBucketResult = {};
+    let key1 = 'Contents';
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: Can you make this a const as I dont see the value of this changing. Also name the const `LIST_CONTENT`

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+    ListBucketResult[key1] = [];
+
+    let commonPrefixes = {};
+    let key2 = 'CommonPrefixes';
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: make it const and name it LIST_COMMON_PREFIX

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+      response.sendStatus(204);
+    }
+  }
+
+  private async handleRequestPut(request, response) {
+    // fill in  this content
+    // console.log("put entered");
+    const directoryPath = join(String(this.localDirectoryPath), String(request.params.path));
+    ensureFileSync(directoryPath);
+    var new_data = stripChunkSignaturev2(request.body);
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: why v2

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+      response.send(o2x({
+        '?xml version="1.0" encoding="utf-8"?': null,
+        InitiateMultipartUploadResult: {
+          "Bucket": this.route,
+          "Key": request.params.path,
+          "UploadId": this.uploadId
+        }
+      }));
+    }
+    if (request.query.uploadId === this.uploadId) {
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: What happens if both conditions fail?

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+        }
+      }));
+    }
+  }
+}
+
+// removing chunk siognature from request payload if present
+function stripChunkSignaturev2(buf: Buffer) {
+  let str = buf.toString();
+  var regex1 = /^[A-Fa-f0-9]+;chunk-signature=[0-9a-f]{64}/gm;
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/S3server.ts]: why regex1

File: packages/amplify-storage-simulator/src/server/index.ts [hunk 20]
```diff
@@ -0,0 +1,27 @@
+import { StorageServer } from './S3server';
```
Comment (yuth) [inline packages/amplify-storage-simulator/src/server/index.ts]: Why do we need this file? Can't we just use S3Server directly?

File: packages/amplify-util-mock/src/storage/storage.ts [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+        const storageParams = context.amplify.readJsonFile(parametersFilePath);
+        this.bucketName = `${storageParams.bucketName}-${localEnvInfo.envName}`
+        const route = path.join(
+            '/',
+            this.bucketName,
+            '/'
+        );
+
+        let localDirS3 = this.createLocalStorage(context, resourceName);
+        const port = 20005; // port for S3
```
Comment (yuth) [inline packages/amplify-util-mock/src/storage/storage.ts]: Move this to the top of the file so its easy to change

File: .gitignore (hunk 1 above)
Comment (akshbhu) [reply .gitignore]: done

File: packages/amplify-frontend-android/lib/frontend-config-creator.js (hunk 2 above)
Comment (akshbhu) [reply packages/amplify-frontend-android/lib/frontend-config-creator.js]: yes, removed it

File: packages/amplify-storage-simulator/package.json (hunk 3 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: changed the author and license

File: packages/amplify-storage-simulator/package.json (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: removed

File: packages/amplify-storage-simulator/package.json (hunk 5 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: yes , I need glob, serve-static , bodyparser but not chokidar

File: packages/amplify-storage-simulator/package.json (hunk 9 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: moved all required @ types to dev dependencies

File: packages/amplify-storage-simulator/src/index.ts (hunk 10 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/index.ts]: removed

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 13 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: removed

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 15 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: done

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 16 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: done

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 17 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: changed the name to stripchunksignature

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 18 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: added a default condition to POST

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 19 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: changed the name to regex. Useful to know the position of signature in buffer

File: packages/amplify-util-mock/src/storage/storage.ts (hunk 21 above)
Comment (akshbhu) [reply packages/amplify-util-mock/src/storage/storage.ts]: done

File: packages/amplify-storage-simulator/src/server/S3server.ts (hunk 14 above)
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: removed

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+      this.server.close();
+      this.server = null;
+      this.connection = null;
+    }
+  }
+
+  private async handleRequestAll(request, response) {
+    // parsing the path and the request parameters
+    request.url = (decodeURIComponent(request.url));
+    var str2 = this.route.slice(0, -1) + '';
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: changed to store the route without '/'.  Now the server can handle both conditions with and without '/'

File: packages/amplify-storage-simulator/src/server/S3server.ts [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+
+    const temp = request.url.split(str2);
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/server/S3server.ts]: getting the path of the object from the URL and storing in the request.params.path and also adding the prefix in front of the path if present.
First, split the URL with the route name, add the rest path to the prefix and store the value in request.params.path

File: .gitignore [hunk 24]
```diff
@@ -6,8 +6,11 @@ coverage
 .nyc_output
 package-lock.json
 packages/amplify-cli/amplify/.config/
+packages/amplify-storage-simulator/public
 .vscode
 packages/graphql*/lib
+packages/amplify-storage-simulator/lib
+packages/amplify-storage-simulator/node_modules
```
Comment (yuth) [inline .gitignore]: This line is not needed

File: packages/amplify-storage-simulator/package.json [hunk 25]
```diff
@@ -0,0 +1,63 @@
+{
+  "name": "amplify-storage-simulator",
+  "version": "0.0.1",
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-storage-simulator/package.json]: changed to 0.1.0

File: packages/amplify-storage-simulator/src/__tests__/S3server.test.ts [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+});
+
+
+describe('Test get api', () => {
+    const actual_file = __dirname + '/test-data/2.png';
+    test('get image work ', async () => {
+        const data = await s3client.getObject({ Bucket: bucket, Key: '2.png'}).promise();
+        expect(data).toBeDefined();
+        expect(data.Body).toBeDefined();
+        //expect(md5(data.Body)).toEqual(md5(actual_file));
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-storage-simulator/src/__tests__/S3server.test.ts]: yes. Still in progress

File: .gitignore (hunk 24 above)
Comment (akshbhu) [reply .gitignore]: removed

**Pull request closed without merging**